packages/editor/src/components/post-format/index.js

```diff
@@ -16,59 +16,73 @@ import { useInstanceId } from '@wordpress/compose';
  */
 import PostFormatCheck from './check';
 
-// All WP post formats, sorted alphabetically by translated name.
-export const POST_FORMATS = [
-	{ id: 'aside', caption: __( 'Aside' ) },
-	{ id: 'audio', caption: __( 'Audio' ) },
-	{ id: 'chat', caption: __( 'Chat' ) },
-	{ id: 'gallery', caption: __( 'Gallery' ) },
-	{ id: 'image', caption: __( 'Image' ) },
-	{ id: 'link', caption: __( 'Link' ) },
-	{ id: 'quote', caption: __( 'Quote' ) },
-	{ id: 'standard', caption: __( 'Standard' ) },
-	{ id: 'status', caption: __( 'Status' ) },
-	{ id: 'video', caption: __( 'Video' ) },
-].sort( ( a, b ) => {
-	const normalizedA = a.caption.toUpperCase();
-	const normalizedB = b.caption.toUpperCase();
-
-	if ( normalizedA < normalizedB ) {
-		return -1;
-	}
-	if ( normalizedA > normalizedB ) {
-		return 1;
-	}
-	return 0;
-} );
+export function usePostFormats() {
+	const { __ } = useTranslate();
+
+	const { supportedFormats } = useSelect( ( select ) => {
+		const { getEditedPostAttribute } = select( 'core/editor' );
+		const _postFormat = getEditedPostAttribute( 'format' );
+		const themeSupports = select( 'core' ).getThemeSupports();
+
+		return {
+			// Ensure current format is always in the set.
+			// The current format may not be a format supported by the theme.
+			supportedFormats: union(
+				[ _postFormat ],
+				get( themeSupports, [ 'formats' ], [] )
+			),
+		};
+	}, [] );
+
+	// All WP post formats, sorted alphabetically by translated name.
+	const formats = [
+		{ id: 'aside', caption: __( 'Aside' ) },
+		{ id: 'audio', caption: __( 'Audio' ) },
+		{ id: 'chat', caption: __( 'Chat' ) },
+		{ id: 'gallery', caption: __( 'Gallery' ) },
+		{ id: 'image', caption: __( 'Image' ) },
+		{ id: 'link', caption: __( 'Link' ) },
+		{ id: 'quote', caption: __( 'Quote' ) },
+		{ id: 'standard', caption: __( 'Standard' ) },
+		{ id: 'status', caption: __( 'Status' ) },
+		{ id: 'video', caption: __( 'Video' ) },
+	].sort( ( a, b ) => {
+		const normalizedA = a.caption.toUpperCase();
+		const normalizedB = b.caption.toUpperCase();
+
+		if ( normalizedA < normalizedB ) {
+			return -1;
+		}
+		if ( normalizedA > normalizedB ) {
+			return 1;
+		}
+		return 0;
+	} );
+
+	return formats.filter( ( format ) =>
+		includes( supportedFormats, format.id )
+	);
+}
```

Comment:
Using `useMemo` would be beneficial here, avoiding translating, sorting and filtering the array over and over on each render:
```js
const formats = useMemo( () => {
  return [ ... ].sort( ... );
}, [ __ ] );

const filteredFormats = useMemo( () => {
  return formats.filter( ... );
}, [ formats, supportedFormats ] );

return filteredFormats;
```